googletest/include/gtest/internal/gtest-internal.h

```diff
@@ -1147,6 +1147,46 @@ class NativeArray {
     GTEST_CONCAT_TOKEN_(gtest_label_testthrow_, __LINE__): \
       fail(gtest_msg.value)
 
+#define GTEST_TEST_THROW_WHAT_(statement, expected_exception, regex, fail) \
+  GTEST_AMBIGUOUS_ELSE_BLOCKER_ \
+  if (::testing::internal::ConstCharPtr gtest_msg = "") { \
+    bool gtest_caught_expected = false; \
+    try { \
+      GTEST_SUPPRESS_UNREACHABLE_CODE_WARNING_BELOW_(statement); \
+    } \
+    catch (expected_exception const& e) { \
+      const ::testing::internal::RE& gtest_regex = (regex); \
+      const bool matched = ::testing::internal::RE::PartialMatch(::testing::PrintToString(e), gtest_regex); \
```

Comment:
On Mon, Nov 30, 2015 at 9:35 PM, Caleb Morse <notifications@github.com>
wrote:

> In googletest/include/gtest/internal/gtest-internal.h
> <https://github.com/google/googletest/pull/641#discussion_r46234204>:
>
> > @@ -1147,6 +1147,46 @@ class NativeArray {
> >      GTEST_CONCAT_TOKEN_(gtest_label_testthrow_, __LINE__): \
> >        fail(gtest_msg.value)
> >
> > +#define GTEST_TEST_THROW_WHAT_(statement, expected_exception, regex, fail) \
> > +  GTEST_AMBIGUOUS_ELSE_BLOCKER_ \
> > +  if (::testing::internal::ConstCharPtr gtest_msg = "") { \
> > +    bool gtest_caught_expected = false; \
> > +    try { \
> > +      GTEST_SUPPRESS_UNREACHABLE_CODE_WARNING_BELOW_(statement); \
> > +    } \
> > +    catch (expected_exception const& e) { \
> > +      const ::testing::internal::RE& gtest_regex = (regex); \
> > +      const bool matched = ::testing::internal::RE::PartialMatch(::testing::PrintToString(e), gtest_regex); \
>
> Would an arbitrary matcher theoretically be able to handle any type of
> exception?
>
The exception would have to be of the type given by the user.
The catcher and the matcher are different roles :)

> To accept a matcher, would I do something like how EXPECT_TRUE/FALSE are
> implemented?
>
> It would be more like EXPECT_THAT.